velox/common/memory/Memory.cpp

```diff
@@ -33,21 +35,44 @@ std::unique_ptr<MemoryManager>& instance() {
   static std::unique_ptr<MemoryManager> kInstance;
   return kInstance;
 }
+
+std::shared_ptr<MemoryAllocator> createAllocator(
+    const MemoryManagerOptions& options) {
+  if (options.allocator != nullptr) {
+    return options.allocator->shared_from_this();
+  }
+  if (options.useMmapAllocator) {
+    MmapAllocator::Options mmapOptions;
+    mmapOptions.capacity =
+        std::min(options.allocatorCapacity, options.capacity);
+    mmapOptions.useMmapArena = options.useMmapArena;
+    mmapOptions.mmapArenaCapacityRatio = options.mmapArenaCapacityRatio;
+    return std::make_shared<MmapAllocator>(mmapOptions);
+  } else {
+    return std::make_shared<MallocAllocator>(
+        std::min(options.allocatorCapacity, options.capacity));
+  }
+}
+
+std::unique_ptr<MemoryArbitrator> createArbitrator(
+    const MemoryManagerOptions& options) {
+  // TODO: consider to reserve a small amount of memory to compensate for the
+  // non-reclaimable cache memory which are pinned by query accesses if enabled.
+  const auto arbitratorCapacity =
+      std::min(options.queryMemoryCapacity, options.arbitratorCapacity);
+  return MemoryArbitrator::create(
+      {.kind = options.arbitratorKind,
+       .capacity = std::min(arbitratorCapacity, options.allocatorCapacity),
+       .memoryPoolTransferCapacity = options.memoryPoolTransferCapacity,
+       .memoryReclaimWaitMs = options.memoryReclaimWaitMs,
+       .arbitrationStateCheckCb = options.arbitrationStateCheckCb});
+}
 } // namespace
 
 MemoryManager::MemoryManager(const MemoryManagerOptions& options)
-    : capacity_{options.capacity},
-      allocator_{options.allocator->shared_from_this()},
+    : allocator_{createAllocator(options)},
```

Comment:
Would like to understand why we want to use a single option to serve both purpose of creating allocator and arbitrator? Would it make sense to have nested option structure to encapsulate each set of options? Thanks

Reply:
It would be better. Then it needs more compatible handling. Consider in followup if feasible. Thanks!